conn.go

```diff
@@ -476,17 +495,61 @@ func (c *Conn) handleData(arg string) {
 	}
 
 	if c.server.LMTP {
-		// TODO: support per-recipient responses
 		for _, rcpt := range c.recipients {
-			c.WriteResponse(code, enhancedCode, "<"+rcpt+"> "+msg)
+			var status *SMTPError
+			rcptStatus := statuscollector.rcptStatus[rcpt]
+			select {
+			case <-rcptStatus.ctx.Done():
+				c.Server().ErrorLog.Printf("Context Error: %s - tempfailing", rcptStatus.ctx.Err())
+				tempfail := SMTPError{
+					Code:         420,
+					EnhancedCode: EnhancedCode{4, 4, 7},
+					Message:      "Error: timeout reached",
+				}
+				status = &tempfail
+			case status = <-rcptStatus.ch:
+			}
+			c.WriteResponse(status.Code, status.EnhancedCode, "<"+rcpt+"> "+status.Message)
 		}
+
 	} else {
 		c.WriteResponse(code, enhancedCode, msg)
 	}
 
 	c.reset()
 }
 
+type rcptStatus struct {
+	ctx context.Context
+	ch  chan *SMTPError
+}
+
+type lmtpStatusCollector struct {
+	wg         sync.WaitGroup
+	rcptStatus map[string]*rcptStatus
+}
+
+func newlmtpStatusCollector(log Logger) *lmtpStatusCollector {
+	return &lmtpStatusCollector{
+		rcptStatus: make(map[string]*rcptStatus),
+	}
+}
+
+func (s *lmtpStatusCollector) SetStatus(rcpt string, status *SMTPError) {
+	defer s.wg.Done()
+	rcpt = strings.ToLower(rcpt)
+	s.rcptStatus[rcpt].ch <- status
+}
+
+func (s *lmtpStatusCollector) StartDelivery(ctx context.Context, rcpt string) {
+	rcpt = strings.ToLower(rcpt)
+	s.rcptStatus[rcpt] = &rcptStatus{
+		ch:  make(chan *SMTPError),
+		ctx: ctx,
+	}
+	s.wg.Add(1)
```

Comment:
I don't get why this is necessary. When `Session.Data` returns the delivery of all receipients is complete. I don't understand why it's necessary to break this assumption.

Reply:
When you are writing a lmtp server which for example stores every rcpt mails to a s3 bucket, you can do this parallel in go routines in background and once it's finished you can reply the code to the client

Reply:
You can just do the wait group stuff in the `Data` command.

Reply:
The wait group is not needed anymore I will remove it

Reply:
You need a way to tell for every rcpt the status at the end of Data. I think the easiest way to solve this it the extended interface. Even with ctx and cancel/timeout support. As lib user you don't care about the logic just use the sc interface, very easy

Reply:
I don't get why `StartDelivery` is necessary at all. You just need to tell go-smtp when a recipient has succeeded/failed.

Reply:
It's needed when you want to reply the SMTPError code immediately to the lmtpclient.

When you wait until all rcpts are finished the client may be timed out and will retry the mail. Lmtp is to prevent this, once it's delivered to a backend the client needs the status immediately

Reply:
Please look into server_test.go how-to use it. Assume there is a http upload in the go routine or maildir delivery

Reply:
>It's needed when you want to reply the SMTPError code immediately to the lmtpclient.

Then just return the error from the function.

I'd rather not change the fact that the delivery is finished exactly when the function returns. It makes everything a lot more complicated.

Reply:
Lmtp is different here the delivery for each rcpt is a own task. If you have many rcpts and a slow delivery to the backend its important to tell tell client about the rcpt status asap. In my case I have such a problem that's why I switched to lmtp instead of smtp

Reply:
Why is it important to tell clients ASAP?

I understand that it's important that the client knows which deliveries have succeeded or failed, but I'm not sure LMTP has been designed for notifying the client as soon as possible. The RFC requires recipient statuses to be sent in order, which somewhat makes that difficult.

Let's say you have 3 recipients, and the first one is very slow. You can't send the status for the completed 2 last recipients before the first one has completed.

Reply:
> Let's say you have 3 recipients, and the first one is very slow. You can't send the status for the completed 2 last recipients before the first one has completed.

Yes thats correct, the final status response is always as fast as the first rcpt delivery, this depends on the ordering in RFC, but we collect the status asap and deliver them to the client

Reply:
Then I'd prefer to just return a value with the status for each recipient.

Reply:
In this case the lib user has todo the ordering correctly and all the channels and context handling, then it would be better to split out lmtp to a own interface.

The other issue is that if you have 50 rcpts and a big mail the delivery can take every long, the lmtp client may times out before we finished all rcpts. In this case the lmtpclient will retry every 50 rcpt (even if there where already 48rcpts successful delivered to the backend).